ext/ddtrace_profiling_native_extension/collectors_cpu_and_wall_time_worker.c

```diff
@@ -0,0 +1,391 @@
+}
+
+// Called in a background thread created in CpuAndWallTimeWorker#start
+static VALUE _native_sampling_loop(DDTRACE_UNUSED VALUE _self, VALUE instance) {
+  struct cpu_and_wall_time_worker_state *state;
+  TypedData_Get_Struct(instance, struct cpu_and_wall_time_worker_state, &cpu_and_wall_time_worker_typed_data, state);
+
+  if (active_sampler_owner_thread != Qnil && is_thread_alive(active_sampler_owner_thread)) {
+    rb_raise(
+      rb_eRuntimeError,
+      "Could not start CpuAndWallTimeWorker: There's already another instance of CpuAndWallTimeWorker active in a different thread"
+    );
+  }
+
+  // This write to a global is thread-safe BECAUSE we're still holding on to the global VM lock at this point
+  active_sampler_instance = instance;
+  active_sampler_owner_thread = rb_thread_current();
+
+  state->should_run = true;
+
+  block_sigprof_signal_handler_from_running_in_current_thread(); // We want to interrupt the thread with the global VM lock, never this one
+
+  install_sigprof_signal_handler(handle_sampling_signal);
+
+  // Release GVL, get to the actual work!
+  int exception_state;
+  rb_protect(release_gvl_and_run_sampling_trigger_loop, instance, &exception_state);
+
+  // The sample trigger loop finished (either cleanly or with an error); let's clean up
+
+  remove_sigprof_signal_handler();
+  active_sampler_instance = Qnil;
+  active_sampler_owner_thread = Qnil;
+
+  // Ensure that instance is not garbage collected while the native sampling loop is running; this is probably not needed, but just in case
+  RB_GC_GUARD(instance);
+
+  if (exception_state) rb_jump_tag(exception_state); // Re-raise any exception that happened
+
+  return Qnil;
+}
+
+static VALUE _native_stop(DDTRACE_UNUSED VALUE _self, VALUE self_instance) {
+  struct cpu_and_wall_time_worker_state *state;
+  TypedData_Get_Struct(self_instance, struct cpu_and_wall_time_worker_state, &cpu_and_wall_time_worker_typed_data, state);
+
+  state->should_run = false;
+
+  return Qtrue;
+}
+
+static void install_sigprof_signal_handler(void (*signal_handler_function)(int, siginfo_t *, void *)) {
+  struct sigaction existing_signal_handler_config = {.sa_sigaction = NULL};
+  struct sigaction signal_handler_config = {
+    .sa_flags = SA_RESTART | SA_SIGINFO,
+    .sa_sigaction = signal_handler_function
+  };
+  sigemptyset(&signal_handler_config.sa_mask);
+
+  if (sigaction(SIGPROF, &signal_handler_config, &existing_signal_handler_config) != 0) {
+    rb_sys_fail("Could not start CpuAndWallTimeWorker: Could not install signal handler");
+  }
+
+  // In some corner cases (e.g. after a fork), our signal handler may still be around, and that's ok
+  if (existing_signal_handler_config.sa_sigaction == handle_sampling_signal) return;
+
+  if (existing_signal_handler_config.sa_handler != NULL || existing_signal_handler_config.sa_sigaction != NULL) {
+    // A previous signal handler already existed. Currently we don't support this situation, so let's just back out
```

Comment:
(For future PR.) You should store this handler somewhere, and make sure to call it at the end of the signal handler. Generally, that's called signal chaining.

Reply:
Yeah... I had heard of that approach, but I was slightly unsure if we shouldn't be a bit more defensive.

E.g., the Ruby VM does not use SIGPROF. So if there's a SIGPROF handler, something else is operating inside the process (perhaps the [stackprof profiler?](https://github.com/tmm1/stackprof/blob/6bbabf14207e8bde5e8d39002f5d1a6f638b0342/ext/stackprof/stackprof.c#L200)).

Since we don't know what's out there, I was thinking it would be safe to just stop and complain, rather than trying to "limp along" by chaining. For instance, perhaps the other profiler has higher overhead, so us sending a SIGPROF every x ms will impact the application because the other profiler hasn't been built to run this often, etc.

Any thoughts on this?